BiometricsFingerprint: uinput: fall back to write on older kernels

krnlyng opened the pull request:
On sailfish XA2 we are currently on 4.4 kernel and this kernel
does not support UI_DEV_SETUP. We need to use the newer fingerprint
HAL since flashing Android 9 with Emma and then SailfishOS currently
breaks fingerprint on SailfishOS. Backporting the fixes for that would
also be an option but staying with upstream vendor-sony-oss-fingerprint
seems better.

So fallback to write if ioctl UI_DEV_SETUP fails in order to avoid
a crashing android.hardware.biometrics.fingerprint service.

Commit:
BiometricsFingerprint: uinput: fall back to write on older kernels

On sailfish XA2 we are currently on 4.4 kernel and this kernel
does not support UI_DEV_SETUP. We need to use the newer fingerprint
HAL since flashing Android 9 with Emma and then SailfishOS currently
breaks fingerprint on SailfishOS. Backporting the fixes for that would
also be an option but staying with upstream vendor-sony-oss-fingerprint
seems better.

So fallback to write if ioctl UI_DEV_SETUP fails in order to avoid
a crashing android.hardware.biometrics.fingerprint service.

MarijnS95 commented on UInput.cpp:
```diff
@@ -60,9 +60,20 @@ int fpc_uinput_create(fpc_uinput_t *uinput)
+
+        struct uinput_user_dev usetup_legacy;
+        memset(&usetup_legacy, 0, sizeof(usetup_legacy));
+        usetup_legacy.id.bustype = BUS_VIRTUAL;
```
A designated initializer is preferred here, in part to stay consistent with the initializer for `struct uinput_setup usetup`.

```suggestion
        struct uinput_user_dev usetup_legacy = {
            .id.bustype = BUS_VIRTUAL,
        };
```

MarijnS95 commented on UInput.cpp:
```diff
@@ -60,9 +60,20 @@ int fpc_uinput_create(fpc_uinput_t *uinput)
+        struct uinput_user_dev usetup_legacy;
+        memset(&usetup_legacy, 0, sizeof(usetup_legacy));
+        usetup_legacy.id.bustype = BUS_VIRTUAL;
+        strcpy(usetup_legacy.name, usetup.name);
```
I'm not too fond of reusing `usetup.name` here, perhaps we can move the device name to a `#define UINPUT_FPC_DEVICE_NAME "uinput-fpc"`?

MarijnS95 commented on UInput.cpp:
```diff
@@ -60,9 +60,20 @@ int fpc_uinput_create(fpc_uinput_t *uinput)
+            ALOGE("Failed to setup uinput device! rc=%d: %s", rc, strerror(errno));
+            close(fd);
+            return rc;
+        }
```
I personally always check whether the full object was written, such as in `fpc_write_input_event`. Though that case should really be improved to print the strerror too, like:

```suggestion
        if (rc < 0) {
            ALOGE("Failed to write legacy uinput setup! rc=%d: %s", rc, strerror(errno));
            close(fd);
            return rc;
        } else if (rc != sizeof(usetup_legacy)) {
            ALOGE("Didn't write full usetup_legacy, only %d/%zu bytes!", rc, sizeof(usetup_legacy));
            /*
            The kernel _should_ always accept the full object, and a loop to write the whole thing is overkill too. shall we just close it and return an error?
            close(fd);
            return -EBUSY;
            */
        }
```
I don't expect this to ever happen though. But better be safe than sorry.

MarijnS95 reviewed (changes requested):
Thanks for the contribution! Fallback tested on k4.14, works as expected.

One stylistic change is requested, the other two are plain paranoia but may be good to have at some point.

krnlyng commented:
I've updated it, i doubt it makes sense to continue if the uinput setup command fails with write so i added the code in your suggestion that was commented.
I've also added another commit addressing devices which have older firmware? At least that's what i think the cause might be.

At least i have one device which was flashed with Emma to android 8 very long ago and then to sailfish which doesn't seem to want to execute the FPC_NAVIGATION_ENTER command properly.

I hope that the same/something similar cannot happen with egistec?

MarijnS95 commented:
@krnlyng Thanks for applying the comments and throwing a `TEMP_FAILURE_RETRY` in there. The first commit is ready to merge IMO.

What device exactly? An XA2 with FPC sensor? I remember seeing similar errors when toying around with Loire and the gesture config (allowing to set things like trigger thresholds) so it might be worth manually pushing such a config to the device. Check the loire-tone-gestures branch on my fork, perhaps you are able to get it to work that way.

I hate re-initializing the device: last I checked `fpc_close` does not properly clean everything up, though it "works fine" after that. I can look into that and make sure it is safe to use, but I'd prefer if you can confirm whether calling `fpc_navi_close` or sending a valid config as detailed above solves the problem.
Be careful about reinitializing the activeGroup too; `AuthenticateAsync()` does this by raising `ERROR_HW_UNAVAILABLE` during a session but gestures are not processed by anything Android is aware of. Besides the active group may or may not have been initialized, as this HAL jumps into gesture processing immediately after starting regardless of additional setup (`setNotify`/`setActiveGroup`) from Android.

Egistec doesn't suffer any of these issues. Only the newer TZapp (found on Ganges, Kumano and updated Niles) have a userspace gesture implementation and seem to work from the get-go. Despite claims that FPC sensors are of higher quality hardware-wise, their TZ-app crashes when not handled with silk gloves and pull of these useless errors _that ruin all other commands too_ :man_shrugging: 

And finally a little style nitpick: I'd rather initialize this simple boolean in-line (`bool mNaviFailed = false;`), the other two pointers are a remnant from [the initial HAL reimplementation](https://github.com/sonyxperiadev/vendor-sony-oss-fingerprint/commit/c35db197a4d44ac4dc777c6b301bb4b1db10f9aa#diff-2052d7532a891a2cdc20d166aced1656R47) which I would move to the header too.
Please limit commit line length (title _and_ description) to 72 chars.

krnlyng commented:
Yes XA2 with FPC sensor.

Unfortunately it seems that i cannot get *any* of the NAGIVATION commands to work, i tried get_config, set_config (to try to set a config before entering the navi state) and exit (like you described, calling it after enter navi failed). All of them make the fingerprint enter the "FPC_ERROR_CONFIG state".

However i came up with a slightly different solution, it's still awkward, and i have NOT tested it on any other device than the one with the old firmware yet. I will do that tomorrow when i have access to another device.

MarijnS95 commented:
I might have misremembered, but the error could have popped up when I tried to use the "newer" code and structures on Suzu. While the structures and numbers are the same, the polling is slightly different. That doesn't explain the error you are seeing though, which occurs immediately when calling `fpc_navi_enter`.

Gestures "work" on Loire and Tone and I don't remember updating Suzu firmware for quite some time. If the TZapp here (iirc it is _not_ on the ODM partition) has support for this feature I don't see why a Nile with old firmware won't be able to support it at all, perhaps something else finicky is going on.

While disabling gestures is the easy way out it turns into a lot of extra management code, in particular the "slightly different solution" that pollutes `fpc_init` with some ugly gotos for this highly unexpected edge-case. I think we can find a way to work around it without requiring a full tzapp reload.

And even then, this is deliberately flashed old firmware. I would personally just request those facing the issue to update it, who knows what other problems older firmware can cause. But on the other hand SFOS is still running 4.4, is this dependent on older firmware too?
Does SFOS support these gestures in the first place? Otherwise we might as well provide a build-var to disable it and call it a day.

For now I suggest the following:
- Remove the navigation edgecase from this PR, leaving only the UInput kernel 4.4 fallback, so that it can be merged;
- Create an issue _or_ new PR specific to this navigation issue. Providing exact firmware/baseband versions so that we can correlate;
- I found some old dumping code and patched it up to extract _some_ information from the binary. While not nearly everything we can pull from it, would you mind to give it a try and report back the results? https://github.com/MarijnS95/vendor-sony-oss-fingerprint/commit/eb3d2117de2803ba534ff8281ff1e4cb71748975.

krnlyng commented:
I realized what the issue is with the continuously failing commands, the return code had to be set to 0 before calling the tz functions, seems it is not automatically set to 0 in case of success and only changed on error?

Doesn't fully make sense to me yet but, it works now without resetting the tz app.

Still would be nice to get those gestures working :).

MarijnS95 commented:
@krnlyng That's really stupid, the TZapp is _supposed_ to overwrite it with the result. But given how bug-riddled it is already this comes as no surprise. After all, simple commands receive the status back as third member of `fpc_send_std_cmd_t`, too.

Would you mind splitting up the patches? One that fixes just this bug (with the appropriate title), one that disables navi (which we hopefully don't need to merge), and one that cleans up the ion buffers which were never freed :man_facepalming: (glad we have scoping and proper destructors in the Egistec codepaths thanks to C++, at least...). is the loire_tone implementation leaking these everywhere too?

> Still would be nice to get those gestures working :).

This makes me curious: Does Sailfish support these gestures too?

Commit:
nile/yoshino/tama: fpc: reset return status

Reset the return status before passing the buffer to the fpc app.

This is because it seems the app doesn't set the return status to 0
when a command succeeds but only sets it on failure.

Commit:
yoshino/nile/tama: free ion buffers on failure

krnlyng commented:
@MarijnS95 they work but in some cases they are not doing what they should ideally do yet i guess.

I've split the patches, i hope it makes sense to you how i split them up.

MarijnS95 commented on fpc_imp_yoshino_nile_tama.c:
```diff
@@ -98,6 +98,7 @@ err_t send_modified_command_to_tz(fpc_data_t *ldata, struct qcom_km_ion_info_t i
     ion_fd_info.data[0].fd = ihandle.ifd_data_fd;
     ion_fd_info.data[0].cmd_buf_offset = 4;
 
+    *(uint32_t*)rec_cmd = 0;
```
How about turning 
```c
void *rec_cmd = handle->ion_sbuffer + TZ_RESPONSE_OFFSET;
```
into
```c
int32_t *rec_cmd = (int32_t *)(handle->ion_sbuffer + TZ_RESPONSE_OFFSET);
```
And using that without cast here and below in `if (result = *rec_cmd)`?

MarijnS95 reviewed (approved):
> I've split the patches, i hope it makes sense to you how i split them up.

Thanks for splitting up the changes in the exact 3 distinct commits, matching exactly what was requested :grinning: 

While I could see a reason to convert the double `ion_free` calls to a `result = -1; goto exit;` that incurs many more unnecessary line changes.

@jerpelea MERGE!

MarijnS95 commented:
@krnlyng Some sensors aren't too precise reading the gestures or require minute movements instead of a full swipe _over_ the sensor. Supporting this is anyway as simple as listening to the generated key events and dealing with it; I hope to reuse left/right swipes and perhaps (double) taps for other navigation besides the notification slide in landscape mode :grin:. When implementing this in SFOS, you guys are aware of the left-right swap in the .idc files to deal with sensors mounted on the back instead of the side of a device?

Commit:
nile/yoshino/tama: fpc: avoid unecessary casting

MarijnS95 commented:
Thanks for cleaning it up :smile: 

@jerpelea MERGE!
